Plugin order

mariusandra opened the pull request:
## Changes

- ~~DRAFT!~~
- Context here: https://github.com/PostHog/plugin-server/issues/102

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

mariusandra committed:
fix form type

mariusandra committed:
split to enabled/disabled plugins and add empty and loading screens

mariusandra committed:
show ingestion order

mariusandra committed:
add "rearrange" button

mariusandra committed:
show alert when rearranging

mariusandra committed:
drag to sort plugins

mariusandra committed:
fixes

mariusandra committed:
move opt-out to "danger zone" in "advanced" tab

mariusandra committed:
refactor "custom" to "advanced" in the code

mariusandra committed:
rename to DangerZone

mariusandra committed:
start rearranging from dragging card

mariusandra committed:
show changed order

mariusandra committed:
remove extra button

mariusandra committed:
save order in the backend

mariusandra committed:
fix default config for newly installed plugins

mariusandra committed:
newly enabled plugins go to the end of the list

mariusandra commented:
This is ready for a first look (some test to fix still?)! Please definitely give feedback on the design and UX. The main bit that changed in this PR is the ordering:

![2021-01-27 10 38 58](https://user-images.githubusercontent.com/53387/105972730-38d36c80-608c-11eb-9440-25709384d428.gif)

Some comments (and replies to https://github.com/PostHog/plugin-server/issues/102): 
- I did in the end split this to two sections: enabled and installed plugins. Playing with it, it seems to work quite well. I think that in situations where you have a lot of installed plugins (e.g. the cloud?), splitting enabled/installed still makes some sense and also makes the ordering logic easier to explain.
- You must confirm the order after rearranging. This is mostly to prevent mistakes, as the accidentally dragging and rearranging the plugins could have consequences
- All newly enabled plugins start out at the end of the list. It's quite easy to move them up.

Other flyby changes with this PR:
- Renamed the tab "Custom" to "Advanced"
- Moved the "opt out of plugins" button into a "danger zone" on the "advanced" tab from the top-right of the page (this prompted the rename of the tab)
- Added an empty state for the installed page with a link to the repository

mariusandra committed:
mypy

yakkomajuri commented:
Immediate thought is that the plugins page is getting a little too crowded - not that massive of a problem at this stage though

Twixes commented on posthog/api/plugin.py:
```diff
@@ -306,3 +306,22 @@ def global_plugins(self, request: request.Request, **kwargs):
+    @action(methods=["PATCH"], detail=False)
+    def rearrange(self, request: request.Request, **kwargs):
+        if not can_configure_plugins_via_api(self.team.organization_id):
+            return Response([])
```
Could maybe be a 403?

mariusandra replied:
makes sense

Twixes commented on posthog/api/plugin.py:
```diff
@@ -306,3 +306,22 @@ def global_plugins(self, request: request.Request, **kwargs):
+        orders = request.data.get("orders", {})
+
+        plugin_configs = PluginConfig.objects.filter(team_id=self.team.pk, enabled=True)
+        plugin_configs_dict = dict((p.plugin_id, p) for p in plugin_configs)
```
Would be a bit nicer with dict comprehension instead of `dict(genexpr)`

Twixes commented on posthog/api/plugin.py:
```diff
@@ -306,3 +306,22 @@ def global_plugins(self, request: request.Request, **kwargs):
+                plugin_config.order = order
+                plugin_config.save()
+
+        reload_plugins_on_workers()
```
Should only reload if plugin_config was saved I think?

Twixes commented on frontend/src/scenes/plugins/utils.ts:
```diff
@@ -39,7 +48,7 @@ export function getPluginConfigFormData(
     const formData = new FormData()
     const otherConfig: Record<string, any> = {}
-    formData.append('enabled', enabled)
+    formData.append('enabled', `${!!enabled}`)
```
Not a super clear construction, how about `Boolean(enabled).toString()`?

Twixes commented on frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx:
```diff
@@ -55,24 +143,86 @@ export function InstalledTab(): JSX.Element {
+                        subtitle={`Installed plugins (${disabledPlugins.length})`}
+                        buttons={<>{enabledPlugins.length === 0 ? upgradeButton : null}</>}
+                    />
+                    <Row gutter={16} style={{ marginTop: 16 }}>
+                        {disabledPlugins.map((plugin) => (
+                            <InstalledPlugin key={plugin.id} plugin={plugin} />
+                        ))}
+                    </Row>
+                </>
+            ) : null}
+
+            {installedPlugins.length === 0 ? (
+                loading ? (
+                    <>
+                        <Subtitle subtitle="Enabled plugins" buttons={<Skeleton.Button style={{ width: 150 }} />} />
+                        <PluginLoading />
+                    </>
+                ) : (
                     <>
-                        {installedPlugins.map((plugin) => {
-                            return <InstalledPlugin key={plugin.id} plugin={plugin} />
-                        })}
-                        {installedPlugins.length == 0 && <Col span={24}>You don't have any plugins installed yet.</Col>}
+                        <Subtitle subtitle="Installed Plugins" />
```
Looks like capitalization differs between Subtitles here.

mariusandra replied:
made the last one lowercase as well

Twixes commented on frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx:
```diff
@@ -37,7 +84,48 @@ export function InstalledTab(): JSX.Element {
+
+        const move = (arr: PluginTypeWithConfig[], from: number, to: number): { id: number; order: number }[] => {
+            const clone = [...arr]
+            Array.prototype.splice.call(clone, to, 0, Array.prototype.splice.call(clone, from, 1)[0])
```
Why on `Array.prototype.splice` instead of just `clone`?

mariusandra replied:
Hmm, not sure I fully understand the question here :). 

We need to clone, since `.splice` alters the array. The `splice` code itself is from a react-sortable-hoc demo and it's the shortest accurate reordering code I've found some far.

Twixes reviewed (commented):
This looks really good, just a few nitpicks.

These arrows are kind of confusing, why are some above and some below?

<img width="47" alt="Screen Shot 2021-01-27 at 11 05 16" src="https://user-images.githubusercontent.com/4550621/105978594-011bf300-6093-11eb-9130-9bdeb3d97e08.png">

As a side note, will this reliably order processing now or is there something to do in the plugin server too?

mariusandra committed:
hide second "check again" button

mariusandra committed:
show code diff

mariusandra committed:
lowercase

mariusandra committed:
clarify string

mariusandra committed:
reload only if saved, add 403 status if no access, simplify dict

mariusandra committed:
set temporary order only on enabled plugins, avoiding rearranging "plugins to update" list

mariusandra commented:
I fixed the nits as well as I could. Not sure about the `clone`/`splice` question though.

Regarding the arrows, The last `3` icon has an arrow above it to show that it's the last one in the queue. Without it there, the "v" below all the other numbers will just look a bit out of place and not leave the impression that it's a queue. Please feel free to suggest something nicer looking! I have two left hands for these things.

I also added one more flyby. Now when updates are available, clicking on the tag for Github and Gitlab plugins will open a diff comparing what exactly changed in the code:
![image](https://user-images.githubusercontent.com/53387/105983687-61ae2e80-6099-11eb-9512-d0f03ec60bd9.png)

mariusandra commented:
And yes, we need a PR for plugin-server as well to use the new order. To be addressed separately...

Twixes commented on frontend/src/scenes/plugins/plugin/PluginCard.tsx:
```diff
@@ -133,11 +134,9 @@ export function PluginCard({
-                                    <CloudDownloadOutlined /> Update available!
-                                </Tag>
-                            ) : latestTag && tag === latestTag ? (
+                            ) : url && latestTag && tag && tag !== latestTag ? (
+                                <UpdateAvailable url={url} tag={tag} latestTag={latestTag} />
+                            ) : url && latestTag && tag && tag === latestTag ? (
```
This looks a bit weird with three conditions duplicated, could they be extracted somehow?

Twixes commented on frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx:
```diff
@@ -158,7 +158,15 @@ export function InstalledTab(): JSX.Element {
+                            ) : pluginsNeedingUpdates.length === 0 ? (
+                                upgradeButton
+                            ) : (
+                                <div />
```
`null` would be more better than an empty element, if possible

mariusandra replied:
That won't work due to some strict type. `<></>` did the trick though

Twixes commented on posthog/api/plugin.py:
```diff
@@ -310,18 +310,21 @@ def global_plugins(self, request: request.Request, **kwargs):
     def rearrange(self, request: request.Request, **kwargs):
         if not can_configure_plugins_via_api(self.team.organization_id):
-            return Response([])
+            return Response(status=403)
```
I may totally be in the wrong, but if we're not returning anything here, `raise PermissionDenied` may work better (error message and all).

mariusandra replied:
I added a `raise ValidationError("Plugin configuration via the web is disabled!")`. Not sure if that's the right error, but this same line was used a few lines higher in the code, and all errors dispatched here are those. Maybe something to check another time. I'll merge it now though if/when the tests pass.

Twixes reviewed (approved):
👍 
A few nitpicks too, but do as you wish about them.

mariusandra committed:
simplify condition

mariusandra committed:
change empty component

mariusandra committed:
raise instead of 403
